### ISSUE TITLE
training cnn with optical flow image

### ISSUE THREAD

**Issue by gaosh** [wadhwasahil/Video-Classification-2-Stream-CNN#10]
Are you training form scratch or finetuning a model pretrained on ImageNet when you training cnn with optical flow images?

**Comment by stillbreeze** [comment]
Optical flow stream has been trained from scratch.

**Comment by fanser** [comment]
@stillbreeze  Has all parameters of  the rgb stream the same learning rate? 
did you fix some layers(conv layers) and fine tune the fc layer  or fine tune all layers with the same lr policy?

**Comment by stillbreeze** [comment]
Yes, the learning rate is the same for all layers and channels. The usual LR decay is used, but it's the same for everything.

The optical flow stream was trained from scratch whereas the the spatial stream was initialised using ImageNet weights and then fine tuned on the CCV dataset.

**Comment by fanser** [comment]
@stillbreeze thanks! 
I notice that your loss fuction is softmax with 20 class, however, the videos of CCV are multi-label, so
1.  It's a good loss function?
2. The label of some video in the train set are all zero ( [0 0 0 ... 0 0 0]), why  do you use the softmax function with 20 class not 21 class (20 pos + 1 neg) ?
Can you tell me the reasons?

**Comment by stillbreeze** [comment]
1. For our experiment, we had an accuracy metric which was considered correct if it could predict any one of the present classes correctly. So, in a sense, we were only doing single-label classification. For predicting multiple classes, maybe something like a per-class binary classifier might work. More sophisticated losses are present too.

2. I can think of one good reason not to use 21 classes with a negative class. The negative class will contain videos which, on their own, don't have any structure as they are effectively random. You don't want your network to pick on some unwanted regularities from these negative classes which will confuse the network when dealing with positive classes. Of course, theres a flip side to all this, when using a normal 20-way softmax to classify a negative class.

**Comment by fanser** [comment]
@stillbreeze 
About the second viewpoint, if use 20-way softmax, it means that the negative classes can't provide any loss for training, beacuse softmax uses the cross-entropy loss and all the negative classes label  are zeros.
it confused me

**Comment by stillbreeze** [comment]
Yes, it's precisely the point of letting the labels be zero. They are ignored. You could've explicitly removed them from the training process. We just chose to let them be there.

For a multi-label problem, you can use a sigmoid instead of a softmax in the last layer and then have a binary classifier for each class.

**Comment by wadhwasahil** [comment]
@fanser After softmax layer, maybe you can put another layer of binary classifier as @stillbreeze just mentioned. Better, just train your network as it is because lets say you softmax output as [0.1, 0.4, 0.25, 0.15, 0.1] and your labels for the video are [0, 1, 1, 0, 0]. Now you could use a cross entropy as it seems apt for the problem or try something different. So I don't know what's the problem here.

**Comment by fanser** [comment]
@wadhwasahil thanks for your reply. 
You are right to use the cross entropy for the multi-labels video.
so the follow question can be removed
> And about the first viewpoint, how to choose one train label from multi-labels during training if use softmax. for example, if a video has labels [1 2 4], which one do you choose from three?
The another question is , for example, that  softmax output as [0.1, 0.4, 0.25, 0.15, 0.1] and your labels for the video are [0, 0,  0, 0, 0]. That's say the negative video can't afford any loss as I said before. And @stillbreeze think we can remove the negative videos  from training process.

@stillbreeze Can you tell me what's your evaluation policy? like MAP, Accuracy or else.Thanks~

**Comment by stillbreeze** [comment]
Yes, you're right about the negative class not contributing anything to the loss, which is why you can remove it. We had already prepared the dataset, so we just let it be.
The evaluation policy was accuracy.

Also, for multiple labels, I'm pretty sure there are better ways to deal with them. Using sigmoid for binary classification instead of softmax is just one of them.

**Comment by fanser** [comment]
@stillbreeze  So when you calculate the accuracy on evaluation set, **do you also remove the negative class from it**? 
20-way softmax couldn't tell us which one is negative, so did you use the **threshold value** to control the prediction. 
thanks for your patience~

**Comment by stillbreeze** [comment]
Yes, you don't account for the negative class when calculating the metric.

Okay, so I will summarise all the options you have:

1. Multi-label classification:
 Use a sigmoid in the last dense layer. Classify each class separately using binary crossentropy loss.
2. Single-label classification:
Use a softmax in the last dense layer. Classify using categorical crossentropy. Choose any one of the ground truth labels for each instance and have only one label per instance. If you don't remove extra classes, then using softmax with categorical crossentropy might not be the most appropriate way, but still works fine in practice.

In all the cases, you can ignore the videos with no labels, or you might choose to use them in an intelligent way that does not interfere with the learning process (some tasks have negative classes as distractors to make the system more generalisable and robust to noise).

**Comment by fanser** [comment]
thanks !
